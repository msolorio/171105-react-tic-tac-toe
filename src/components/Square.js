import React from 'react';
import PropTypes from 'prop-types';

export default function Square(props) {

  function handleSquareClick() {
    props.onSquareClick(props.index);
  }

  return (
    <div className={`Square Square-${props.index}`}
      onClick={handleSquareClick}>
      {props.squareVal}
    </div>
  );
}

Square.propTypes = {
  index: PropTypes.number.isRequired,
  squareVal: PropTypes.string.isRequired,
  onSquareClick: PropTypes.func.isRequired
};
