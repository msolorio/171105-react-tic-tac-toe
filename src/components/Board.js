import React from 'react';
import PropTypes from 'prop-types';
import Square from './Square';

export default function Board(props) {

  function renderSquares() {
    return props.squareVals.map((squareVal, index) => {

      return (
        <Square index={index}
          squareVal={squareVal}
          onSquareClick={props.onSquareClick}
          key={index} />
      );
    });
  }

  return <div className="Board">{renderSquares()}</div>;
}

Board.propTypes = {
  squareVals: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSquareClick: PropTypes.func.isRequired
};
