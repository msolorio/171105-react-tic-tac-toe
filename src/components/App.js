import React, { Component } from 'react';
import Board from './Board';
import Message from './Message';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      squareVals: Array(9).fill(null),
      xIsNext: true
    };

    this.handleSquareClick = this.handleSquareClick.bind(this);
  }

  calculateWinner() {
    const winningCombos = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 4, 8],
      [6, 4, 2]
    ];

    const squareVals = this.state.squareVals;

    return winningCombos.reduce((winner, combo) => {
      const [a, b, c] = combo;
      if (
        squareVals[a] !== null
        && squareVals[a] === squareVals[b]
        && squareVals[a] === squareVals[c]
      ) {
        return squareVals[a];
      }
      return winner;
    }, null);
  }

  handleSquareClick(index) {
    if (this.calculateWinner()) return;
    if (this.state.squareVals[index] !== null) return;
    const squareValsClone = this.state.squareVals.slice();
    squareValsClone[index] = this.state.xIsNext ? 'X' : 'O';

    this.setState((prevState) => ({
      squareVals: squareValsClone,
      xIsNext: !prevState.xIsNext
    }));
  }

  render() {
    return (
      <div className="App">
        <Message winner={this.calculateWinner()} xIsNext={this.state.xIsNext} />
        <Board squareVals={this.state.squareVals}
          onSquareClick={this.handleSquareClick} />
      </div>
    );
  }
}

export default App;
