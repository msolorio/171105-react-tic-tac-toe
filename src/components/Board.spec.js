import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import App from './App';
import Board from './Board';
import Square from './Square';

const boardProps = {
  squareVals: Array(9).fill(null),
  onSquareClick: jest.fn()
};

describe('Board', () => {
  it('should match snapshot', () => {
    const tree = renderer.create(
      <Board {...boardProps} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should pass onSquareClick as a prop to Square', () => {
    const firstSquare = mount(<Board {...boardProps} />).find(Square).first();

    expect(firstSquare.prop('onSquareClick')).toBe(boardProps.onSquareClick);
  });

  it('should pass the correct square value to a squares based on squareVals prop passed in', () => {
    const firstSquare = mount(<Board {...boardProps} />).find(Square).first();
    expect(firstSquare.prop('squareVal')).toBe(null);

    const boardPropsFilled = {
      squareVals: ['X', 'O', null, null, null, null, null, null, null],
      onSquareClick: jest.fn()
    };

    const wrapper = mount(<Board {...boardPropsFilled} />);
    const firstSquareFilled = wrapper.find(Square).first();
    const secondSquareFilled = wrapper.find(Square).at(1);
    const thirdSquareFilled = wrapper.find(Square).at(2);

    expect(firstSquareFilled.prop('squareVal')).toBe('X');
    expect(secondSquareFilled.prop('squareVal')).toBe('O');
    expect(thirdSquareFilled.prop('squareVal')).toBe(null);
  });

  it('should render 9 squares', () => {
    const wrapper = mount(<Board {...boardProps} />);
    const squares = wrapper.find(Square);
    
    expect(squares).toHaveLength(9);
  });
});
