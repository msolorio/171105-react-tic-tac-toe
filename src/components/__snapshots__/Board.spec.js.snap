// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should match snapshot 1`] = `
<div
  className="Board"
>
  <div
    className="Square Square-0"
    onClick={[Function]}
  />
  <div
    className="Square Square-1"
    onClick={[Function]}
  />
  <div
    className="Square Square-2"
    onClick={[Function]}
  />
  <div
    className="Square Square-3"
    onClick={[Function]}
  />
  <div
    className="Square Square-4"
    onClick={[Function]}
  />
  <div
    className="Square Square-5"
    onClick={[Function]}
  />
  <div
    className="Square Square-6"
    onClick={[Function]}
  />
  <div
    className="Square Square-7"
    onClick={[Function]}
  />
  <div
    className="Square Square-8"
    onClick={[Function]}
  />
</div>
`;
